src/coreclr/jit/llvm.cpp

```diff
@@ -1424,6 +1444,119 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
+{
+    Value* paddingDestAddress = _builder.CreateGEP(baseAddress, _builder.getInt32(startOffset));
+
+    const unsigned paddingSize = endOffset - startOffset;
+    // copy the bytes in the padding
+    _builder.CreateMemCpy(paddingDestAddress, llvm::Align(), srcAddress, llvm::Align(), paddingSize);
+
+    return paddingSize;
+}
+
+void Llvm::storeObjAtAddress(Value* baseAddress, Value* data, StructDesc* structDesc)
+{
+    unsigned fieldCount = structDesc->getFieldCount();
+    unsigned bytesStored = 0;
+
+    for (unsigned i = 0; i < fieldCount; i++)
+    {
+        FieldDesc* fieldDesc = structDesc->getFieldDesc(i);
+        unsigned   fieldOffset = fieldDesc->getFieldOffset();
+        Value*     address     = _builder.CreateGEP(baseAddress, _builder.getInt32(fieldOffset));
+
+        if (structDesc->isExplicitLayout() && fieldOffset > bytesStored)
+        {
+            bytesStored += buildMemCpy(baseAddress, bytesStored, fieldOffset, address);
+        }
+
+        Value* fieldData = nullptr;
+        if (data->getType()->isStructTy())
+        {
+            const llvm::StructLayout* structLayout = _module->getDataLayout().getStructLayout(static_cast<llvm::StructType*>(data->getType()));
+
+            unsigned llvmFieldIndex = structLayout->getElementContainingOffset(fieldOffset);
+            fieldData               = _builder.CreateExtractValue(data, llvmFieldIndex);
+        }
+        else
+        {
+            // single field IL structs are not LLVM structs
+            fieldData = data;
+        }
+
+        if (fieldData->getType()->isStructTy())
+        {
+            assert(fieldDesc->getClassHandle() != NO_CLASS_HANDLE);
+
+            // recurse into struct
+            storeObjAtAddress(address, fieldData, getStructDesc(fieldDesc->getClassHandle()));
+
+            bytesStored += fieldData->getType()->getPrimitiveSizeInBits() / 8;
+        }
+        else
+        {
+            if (fieldDesc->isGcPointer())
+            {
+                _builder.CreateCall(getOrCreateRhpAssignRef(),
+                                    ArrayRef<Value*>{address,
+                                                     castIfNecessary(fieldData, Type::getInt8PtrTy(_llvmContext))});
+                bytesStored += TARGET_POINTER_SIZE;
+            }
+            else
+            {
+                _builder.CreateStore(fieldData, castIfNecessary(address, fieldData->getType()->getPointerTo()));
+
+                bytesStored += fieldData->getType()->getPrimitiveSizeInBits() / 8;
+            }
+        }
+    }
+
+    unsigned llvmStructSize = data->getType()->getPrimitiveSizeInBits() / 8;
+    if (structDesc->isExplicitLayout() && llvmStructSize > bytesStored)
+    {
+        Value* srcAddress = _builder.CreateGEP(baseAddress, _builder.getInt32(bytesStored));
+
+        buildMemCpy(baseAddress, bytesStored, llvmStructSize, srcAddress);
+    }
+}
+
+void Llvm::buildStoreObj(GenTreeIndir* indirOp)
```

Comment:
The method is named `buildStoreObj` but takes `GenTreeIndir*`. Why not make it take `GenTreeObj*`?

(Also I would rename `indirOp` to something with `store` in the name)

I think we need to establish a naming convention for trees-vs-LLVM-values, they are mixed right now and it can be confusing. Appending `node` for trees seems a fine solution, e. g. here we could have:
```c
GenTree* dstAddrNode = storeObjNode->Addr();
Value* baseAddress = getGenTreeValue(dstAddrNode);
```
Or we could leave the trees with "simple" names and append `Value` to `llvm::Value*`s instead, e. g.:
```
GenTree* dstAddr = storeObj->Addr();
Value* baseAddressValue = getGenTreeValue(dstAddr);
```
What do you think?

Reply:
@yowl ^